tools/trace_analyzer_tool.cc

```diff
@@ -0,0 +1,1621 @@
+            stats.top_k_queue.push(
+                std::make_pair(it->second.access_count, it->first));
+          } else {
+            if (it->second.access_count > stats.top_k_queue.top().first) {
+              stats.top_k_queue.pop();
+              stats.top_k_queue.push(
+                  std::make_pair(it->second.access_count, it->first));
+            }
+          }
+        }
+      }
+    }
+  }
+  return Status::OK();
+}
+
+
+
+// End the processing, print the requested results
+Status TraceAnalyzer::EndProcessing() {
+  if (trace_sequence_f != nullptr) {
+    fclose(trace_sequence_f);
+  }
+  PrintGetStatistics();
+  CloseOutputFiles();
+  return Status::OK();
+}
+
+// Insert the corresponding key statistics to the correct type
+// and correct CF, output the time-serial file if needed
+Status TraceAnalyzer::KeyStatsInsertion(const uint32_t& type,
+                                        const uint32_t& cf_id,
+                                        const std::string& key,
+                                        const size_t value_size,
+                                        const uint64_t ts) {
+  auto found_stats = ta_[type].stats.find(cf_id);
+  Status s;
+  StatsUnit unit;
+  unit.key_id = 0;
+  unit.cf_id = cf_id;
+  unit.value_size = value_size;
+  unit.access_count = 1;
+  unit.latest_ts = ts;
+  unit.latest_type = type;
+  unit.v_corre.resize(analyzer_opts_.corre_list.size());
+  std::string prefix;
+  if (analyzer_opts_.output_prefix_cut) {
+    prefix = key.substr(0, analyzer_opts_.prefix_cut);
+  }
+
+  if (begin_time_ == 0) {
+    begin_time_ = ts;
+  }
+  uint32_t time_in_sec;
+  if (ts < begin_time_) {
+    time_in_sec = 0;
+  } else {
+    time_in_sec = (ts - begin_time_)/1000000;
+  }
+  uint64_t dist_value_size = value_size/analyzer_opts_.value_interval;
+  if (found_stats == ta_[type].stats.end()) {
+    TraceStats new_stats;
+    new_stats.cf_id = cf_id;
+    new_stats.cf_name = std::to_string(cf_id);
+    new_stats.a_count = 1;
+    new_stats.akey_id = 0;
+    new_stats.a_key_size_sqsum = key.size()*key.size();
+    new_stats.a_key_size_sum = key.size();
+    new_stats.a_value_size_sqsum = value_size*value_size;
+    new_stats.a_value_size_sum = value_size;
+    s = OpenStatsOutputFiles(ta_[type].type_name, new_stats);
+    new_stats.a_key_stats[key] = unit;
+    new_stats.a_value_size_stats[dist_value_size] = 1;
+    new_stats.a_io_stats[time_in_sec] = 1;
+    new_stats.corre_output.resize(analyzer_opts_.corre_list.size());
+    if (analyzer_opts_.output_prefix_cut) {
+      std::map<std::string, uint32_t> tmp_io_map;
+      tmp_io_map[prefix] = 1;
+      new_stats.a_io_prefix_stats[time_in_sec] = tmp_io_map;
+    }
+    ta_[type].stats[cf_id] = new_stats;
+  } else {
+    found_stats->second.a_count++;
+    found_stats->second.a_key_size_sqsum += key.size()*key.size();
+    found_stats->second.a_key_size_sum += key.size();
+    found_stats->second.a_value_size_sqsum += value_size*value_size;
```

Comment:
This may overflow if user has length of GBs. You probably should check overflow here. It's OK to just exit saying it's not supported.

Reply:
Sure